lib/pbench/server/api/resources/datasets_list.py

```diff
@@ -331,22 +332,45 @@ def keyspace(self, query: Query) -> JSONOBJECT:
         Returns:
             The aggregated keyspace JSON object
         """
-        aggregate: JSONOBJECT = {
-            "dataset": {c.name: None for c in Dataset.__table__._columns}
-        }
-
         Database.dump_query(query, current_app.logger)
+        aggregate: JSONOBJECT = {}
 
         datasets = query.all()
         for d in datasets:
+            if not aggregate:
+                aggregate.update(
+                    {"dataset": {c.name: None for c in Dataset.__table__._columns}}
+                )
             for m in d.metadatas:
                 # "metalog" is a top-level key in the Metadata schema, but we
                 # report it as a sub-key of "dataset".
                 if m.key == Metadata.METALOG:
                     self.accumulate(aggregate["dataset"], m.key, m.value)
                 else:
                     self.accumulate(aggregate, m.key, m.value)
-        return aggregate
+        return {"keys": aggregate}
```

Comment:
> I can't help tinkering with this (having to do the not aggregate check for every dataset is dissatisfying...). Perhaps this will suit?

Ugh. So we're going to compare two dicts to see whether to throw it away? The "is-empty" check is relatively simple, and I don't think *this* is an improvement.

(It wouldn't work anyway without doing a *deep* copy of `base_aggregate` since otherwise you're just aliasing the collection and they'll always be the same; and the deep copy is yet another unnecessary expense.)

> It also occurs to me that, since we're constructing an object to return at the end anyway, perhaps we can move the other `if` statement outside the loop:

We'd have to skip the `metalog` key in the first loop, or else waste time computing it and then `del aggregate["metalog"]` at the end. We don't want both `metalog` and `dataset.metalog`. I'd rather keep it as-is.

> And, if we can assume that `datasets` is non-empty iff `Metadata.METALOG in aggregate` is true, then we can combine the ternary in the `return` statement the `if Metadata.METALOG` statement. (Is that correct?)

I'm not even sure what you mean here; but ultimately I'd rather leave it alone!

Reply:
> we're going to compare two dicts to see whether to throw it away? The "is-empty" check is relatively simple, and I don't think _this_ is an improvement.

The dict-dict comparison fails quickly if they are different, and, if they are not different, then the two dicts are small and straightforward to compare -- so either way it's a limited expense; on the other side of the balance is the "is-empty" check which is a very small incremental cost but it will build up if we have a large number of datasets

> It wouldn't work anyway without doing a _deep_ copy of `base_aggregate` since otherwise you're just aliasing the collection and they'll always be the same

That's a good point.  However, instead of capturing `base_aggregate`, we could just compare `aggregate` to the original comprehension again.

> and the deep copy is yet another unnecessary expense

As I said, it think the initial aggregation value is small, and I expect that it would be easy to regenerate; the question is, how expensive is the alternative (e.g., what if there are lots of datasets?).

> We'd have to skip the `metalog` key in the first loop, or else waste time computing it and then `del aggregate["metalog"]` at the end. We don't want both `metalog` and `dataset.metalog`.

I don't follow.  I'm proposing that we generate the subtree for `metalog` where it naturally falls, and then, when it's all done, move it to where we want it.  Adding it as `dataset.metalog` is a cheap operation (it's just a reference copy), and then removing the top-level `metalog` key is cheap as well.  (It's not quite as _pretty_ as I would like, but it's not expensive.)  I don't think we have to skip anything, and the cost of moving at the end is offset by not having to check each key for `"metalog"`.

> > And, if we can assume that `datasets` is non-empty iff `Metadata.METALOG in aggregate` is true, then we can combine the ternary in the `return` statement the `if Metadata.METALOG` statement. (Is that correct?)
> 
> I'm not even sure what you mean here; but ultimately I'd rather leave it alone!

I was wondering whether we could detect an empty datasets list by finding no `metalog` key in the aggregate, but I don't know whether that approach hangs together.

But, since you would prefer to leave it alone, I shall resolve this conversation.  🙂